docs/eBpfExtensions.md

```diff
@@ -189,7 +199,17 @@ This optional function is used to populate the flat buffers representing the dat
 returned to the application when the `bpf_prog_test_run_opts` call completes. In addition, the function frees any
 resources allocated in the `ebpf_program_context_create_t` call.
 
-### 2.2 Program Information NPI Client Attach and Detach Callbacks
+### 2.2 Backward compatibility of the Extension data structures
+All the extension data structures are versioned. New fields can be added to the end of a data structure to maintain backward compatibility
+with the existing extensions. In such cases, the size field of the header will be updated but the version of the structure *will not change*.
+Existing eBPF extensions will continue to work without requiring recompilation.
```

Comment:
In such a case, I think we should add explicit padding in all these structs to ensure compiler does not automatically add padding and increase the size? (and also enabled compiler warning `C4820`)?

Reply:
I would recommend against explicit padding at the end since it would unduly bloat the structure.  For example, say you have 1 byte of non-header data in a struct at present, and later want to add a second 1 byte field.  If it's padded already then you cannot reuse that space for the new field since you have to add more space to get the size to increase.  If on the other hand the struct is packed (no padding) then you can easily add just 1 more byte and all is well.

Reply:
The proposal in this doc is that if new fields are to be added, the version need not be incremented, and just the size of the struct should be sufficient to determine whether the struct is of older or newer "version". 

This logic will break if compiler had earlier added padding and adding a new field _does not_ increase the struct size (as the new field got absorbed in the padding). In such a case, just the struct size may not be sufficient to determine if there is an extra field in the struct.

Reply:
Use the [`pack` pragma](https://learn.microsoft.com/en-us/cpp/preprocessor/pack?view=msvc-170) to guarantee that it won't be padded.

Reply:
One minor concern with using `pack` pragma is that the fields may not be aligned, and can possibly cause performance issues, but I think in our case NMR interface objects will be in hot path.

Reply:
@dthaler - that is why I think having a minor version in addition the length field will be useful. And we could maintain an expected length per version. Such as:

`#define EBPF_HELPER_FUNCTION_PROTOTYPE_VERSION_1_2_EXPECTED_LENGTH offsetof(ebpf_helper_function_prototype_t, the_new_field`) + sizeof(typeof(the_new_field))`.

If you are OK with this, I can publish another iteration of the changes. If not, I will pack to 1 byte which may have performance issues.

Reply:
I suspect the minor version concept is more bug prone, and we didn't go that route for NMR.  Just using sizeof(the_struct) seems less bug prone to me.  I'd rather either pack to 1 byte, or use explicit padding as Anurag suggested.  (And you can use explicit internal padding to ensure alignment if you do go the pack route.)  But I think Anurag's suggestion would be preferable to having a minor version.   That's my 2 cents.

Reply:
I chose to move ahead with the minor version and no packing.